File: cathode/src/main/java/net/simonvt/cathode/ui/show/ShowViewModel.java
/*
 * Copyright (C) 2018 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.ui.show;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import java.util.List;
import net.simonvt.cathode.common.data.MappedCursorLiveData;
import net.simonvt.cathode.common.data.StringMapper;
import net.simonvt.cathode.common.entity.CastMember;
import net.simonvt.cathode.common.entity.Comment;
import net.simonvt.cathode.common.entity.Episode;
import net.simonvt.cathode.common.entity.Season;
import net.simonvt.cathode.common.entity.Show;
import net.simonvt.cathode.entitymapper.CommentListMapper;
import net.simonvt.cathode.entitymapper.EpisodeMapper;
import net.simonvt.cathode.entitymapper.SeasonListMapper;
import net.simonvt.cathode.entitymapper.ShowCastMapper;
import net.simonvt.cathode.entitymapper.ShowListMapper;
import net.simonvt.cathode.entitymapper.ShowMapper;
import net.simonvt.cathode.provider.DatabaseContract.CommentColumns;
import net.simonvt.cathode.provider.DatabaseContract.EpisodeColumns;
import net.simonvt.cathode.provider.DatabaseContract.PersonColumns;
import net.simonvt.cathode.provider.DatabaseContract.RelatedShowsColumns;
import net.simonvt.cathode.provider.DatabaseContract.SeasonColumns;
import net.simonvt.cathode.provider.DatabaseContract.ShowCastColumns;
import net.simonvt.cathode.provider.DatabaseContract.ShowColumns;
import net.simonvt.cathode.provider.DatabaseContract.ShowGenreColumns;
import net.simonvt.cathode.provider.DatabaseSchematic.Tables;
import net.simonvt.cathode.provider.ProviderSchematic.Comments;
import net.simonvt.cathode.provider.ProviderSchematic.Episodes;
import net.simonvt.cathode.provider.ProviderSchematic.RelatedShows;
import net.simonvt.cathode.provider.ProviderSchematic.Seasons;
import net.simonvt.cathode.provider.ProviderSchematic.ShowCast;
import net.simonvt.cathode.provider.ProviderSchematic.ShowGenres;
import net.simonvt.cathode.provider.ProviderSchematic.Shows;
import net.simonvt.cathode.provider.util.SqlColumn;

public class ShowViewModel extends AndroidViewModel {

  static final String[] SHOW_PROJECTION = new String[] {
      ShowColumns.ID, ShowColumns.TRAKT_ID, ShowColumns.TITLE, ShowColumns.YEAR,
      ShowColumns.AIR_TIME, ShowColumns.AIR_DAY, ShowColumns.NETWORK, ShowColumns.CERTIFICATION,
      ShowColumns.STATUS, ShowColumns.USER_RATING, ShowColumns.RATING, ShowColumns.OVERVIEW,
      ShowColumns.IN_WATCHLIST, ShowColumns.IN_COLLECTION_COUNT, ShowColumns.WATCHED_COUNT,
      ShowColumns.LAST_SYNC, ShowColumns.LAST_COMMENT_SYNC, ShowColumns.LAST_CREDITS_SYNC,
      ShowColumns.LAST_RELATED_SYNC, ShowColumns.HOMEPAGE, ShowColumns.TRAILER, ShowColumns.IMDB_ID,
      ShowColumns.TVDB_ID, ShowColumns.TMDB_ID, ShowColumns.NEEDS_SYNC, ShowColumns.HIDDEN_CALENDAR,
  };

  public static final String[] SEASONS_PROJECTION = new String[] {
      SeasonColumns.ID, SeasonColumns.SHOW_ID, SeasonColumns.SEASON, SeasonColumns.UNAIRED_COUNT,
      SeasonColumns.WATCHED_COUNT, SeasonColumns.IN_COLLECTION_COUNT, SeasonColumns.AIRED_COUNT,
      SeasonColumns.WATCHED_AIRED_COUNT, SeasonColumns.COLLECTED_AIRED_COUNT,
      SeasonColumns.LAST_MODIFIED, SeasonColumns.SHOW_TITLE,
  };

  static final String[] EPISODE_PROJECTION = new String[] {
      EpisodeColumns.ID, EpisodeColumns.TITLE, EpisodeColumns.FIRST_AIRED, EpisodeColumns.SEASON,
      EpisodeColumns.EPISODE, EpisodeColumns.WATCHED, EpisodeColumns.WATCHING,
      EpisodeColumns.CHECKED_IN,
  };

  static final String[] RELATED_PROJECTION = new String[] {
      SqlColumn.table(Tables.SHOW_RELATED).column(RelatedShowsColumns.RELATED_SHOW_ID),
      SqlColumn.table(Tables.SHOWS).column(ShowColumns.ID),
      SqlColumn.table(Tables.SHOWS).column(ShowColumns.TITLE),
      SqlColumn.table(Tables.SHOWS).column(ShowColumns.OVERVIEW),
      SqlColumn.table(Tables.SHOWS).column(ShowColumns.RATING),
      SqlColumn.table(Tables.SHOWS).column(ShowColumns.VOTES),
  };

  static final String[] GENRES_PROJECTION = new String[] {
      ShowGenreColumns.GENRE,
  };

  private long showId = -1L;

  private LiveData<Show> show;
  private LiveData<List<String>> genres;
  private LiveData<List<Season>> seasons;
  private LiveData<List<CastMember>> cast;
  private LiveData<List<Comment>> userComments;
  private LiveData<List<Comment>> comments;
  private LiveData<List<Show>> related;
  private LiveData<Episode> toWatch;
  private LiveData<Episode> lastWatched;
  private LiveData<Episode> toCollect;
  private LiveData<Episode> lastCollected;

  public ShowViewModel(@NonNull Application application) {
    super(application);
  }

  public void setShowId(long showId) {
    if (this.showId == -1L) {
      this.showId = showId;

      show =
          new MappedCursorLiveData<>(getApplication(), Shows.withId(showId), SHOW_PROJECTION, null,
              null, null, new ShowMapper());
      genres = new MappedCursorLiveData<>(getApplication(), ShowGenres.fromShow(showId),
          GENRES_PROJECTION, null, null, null, new StringMapper(ShowGenreColumns.GENRE));
      seasons =
          new MappedCursorLiveData<>(getApplication(), Seasons.fromShow(showId), SEASONS_PROJECTION,
              null, null, Seasons.DEFAULT_SORT, new SeasonListMapper());
      cast = new MappedCursorLiveData<>(getApplication(), ShowCast.fromShow(showId),
          ShowCastMapper.PROJECTION, Tables.PEOPLE + "." + PersonColumns.NEEDS_SYNC + "=0", null,
          Tables.SHOW_CAST + "." + ShowCastColumns.ID + " ASC LIMIT 3", new ShowCastMapper());
      userComments = new MappedCursorLiveData<>(getApplication(), Comments.fromShow(showId),
          CommentListMapper.PROJECTION, CommentColumns.IS_USER_COMMENT + "=1", null, null,
          new CommentListMapper());
      comments = new MappedCursorLiveData<>(getApplication(), Comments.fromShow(showId),
          CommentListMapper.PROJECTION,
          CommentColumns.IS_USER_COMMENT + "=0 AND " + CommentColumns.SPOILER + "=0", null,
          CommentColumns.LIKES + " DESC LIMIT 3", new CommentListMapper());
      related = new MappedCursorLiveData<>(getApplication(), RelatedShows.fromShow(showId),
          RELATED_PROJECTION, null, null, RelatedShowsColumns.RELATED_INDEX + " ASC LIMIT 3",
          new ShowListMapper());
      toWatch = new WatchedLiveData(getApplication(), showId, EPISODE_PROJECTION);
      lastWatched = new MappedCursorLiveData<>(getApplication(), Episodes.fromShow(showId),
          EPISODE_PROJECTION, EpisodeColumns.WATCHED + "=1", null,
          EpisodeColumns.SEASON + " DESC, " + EpisodeColumns.EPISODE + " DESC LIMIT 1",
          new EpisodeMapper());
      toCollect = new MappedCursorLiveData<>(getApplication(), Episodes.fromShow(showId),
          EPISODE_PROJECTION, EpisodeColumns.IN_COLLECTION
          + "=0 AND "
          + EpisodeColumns.FIRST_AIRED
          + " IS NOT NULL AND "
          + EpisodeColumns.SEASON
          + ">0", null, EpisodeColumns.SEASON + " ASC, " + EpisodeColumns.EPISODE + " ASC LIMIT 1",
          new EpisodeMapper());
      lastCollected = new MappedCursorLiveData<>(getApplication(), Episodes.fromShow(showId),
          EPISODE_PROJECTION, EpisodeColumns.IN_COLLECTION + "=1", null,
          EpisodeColumns.SEASON + " DESC, " + EpisodeColumns.EPISODE + " DESC LIMIT 1",
          new EpisodeMapper());
    }
  }

  public LiveData<Show> getShow() {
    return show;
  }

  public LiveData<List<String>> getGenres() {
    return genres;
  }

  public LiveData<List<Season>> getSeasons() {
    return seasons;
  }

  public LiveData<List<CastMember>> getCast() {
    return cast;
  }

  public LiveData<List<Comment>> getUserComments() {
    return userComments;
  }

  public LiveData<List<Comment>> getComments() {
    return comments;
  }

  public LiveData<List<Show>> getRelated() {
    return related;
  }

  public LiveData<Episode> getToWatch() {
    return toWatch;
  }

  public LiveData<Episode> getLastWatched() {
    return lastWatched;
  }

  public LiveData<Episode> getToCollect() {
    return toCollect;
  }

  public LiveData<Episode> getLastCollected() {
    return lastCollected;
  }
}

File: cathode/src/main/java/net/simonvt/cathode/ui/show/ShowViewModel.kt
/*
 * Copyright (C) 2018 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.ui.show

import android.content.Context
import android.text.format.DateUtils
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.launch
import net.simonvt.cathode.actions.comments.SyncShowComments
import net.simonvt.cathode.actions.invokeAsync
import net.simonvt.cathode.actions.shows.SyncRelatedShows
import net.simonvt.cathode.actions.shows.SyncShow
import net.simonvt.cathode.actions.shows.SyncShowCollectedStatus
import net.simonvt.cathode.actions.shows.SyncShowCredits
import net.simonvt.cathode.actions.shows.SyncShowImages
import net.simonvt.cathode.actions.shows.SyncShowWatchedStatus
import net.simonvt.cathode.common.data.MappedCursorLiveData
import net.simonvt.cathode.common.data.StringMapper
import net.simonvt.cathode.entity.CastMember
import net.simonvt.cathode.entity.Comment
import net.simonvt.cathode.entity.Episode
import net.simonvt.cathode.entity.Season
import net.simonvt.cathode.entity.Show
import net.simonvt.cathode.entitymapper.CommentListMapper
import net.simonvt.cathode.entitymapper.CommentMapper
import net.simonvt.cathode.entitymapper.EpisodeMapper
import net.simonvt.cathode.entitymapper.SeasonListMapper
import net.simonvt.cathode.entitymapper.SeasonMapper
import net.simonvt.cathode.entitymapper.ShowCastMapper
import net.simonvt.cathode.entitymapper.ShowListMapper
import net.simonvt.cathode.entitymapper.ShowMapper
import net.simonvt.cathode.provider.DatabaseContract.CommentColumns
import net.simonvt.cathode.provider.DatabaseContract.EpisodeColumns
import net.simonvt.cathode.provider.DatabaseContract.RelatedShowsColumns
import net.simonvt.cathode.provider.DatabaseContract.ShowCastColumns
import net.simonvt.cathode.provider.DatabaseContract.ShowGenreColumns
import net.simonvt.cathode.provider.DatabaseSchematic.Tables
import net.simonvt.cathode.provider.ProviderSchematic.Comments
import net.simonvt.cathode.provider.ProviderSchematic.Episodes
import net.simonvt.cathode.provider.ProviderSchematic.RelatedShows
import net.simonvt.cathode.provider.ProviderSchematic.Seasons
import net.simonvt.cathode.provider.ProviderSchematic.ShowCast
import net.simonvt.cathode.provider.ProviderSchematic.ShowGenres
import net.simonvt.cathode.provider.ProviderSchematic.Shows
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.settings.TraktLinkSettings
import net.simonvt.cathode.ui.RefreshableViewModel
import javax.inject.Inject

class ShowViewModel @Inject constructor(
  private val context: Context,
  private val showHelper: ShowDatabaseHelper,
  private val syncShow: SyncShow,
  private val syncRelatedShows: SyncRelatedShows,
  private val syncShowImages: SyncShowImages,
  private val syncShowComments: SyncShowComments,
  private val syncShowCredits: SyncShowCredits,
  private val syncShowWatchedStatus: SyncShowWatchedStatus,
  private val syncShowCollectedStatus: SyncShowCollectedStatus
) : RefreshableViewModel() {

  private var showId = -1L

  lateinit var show: LiveData<Show>
    private set
  lateinit var genres: LiveData<List<String>>
    private set
  lateinit var seasons: LiveData<List<Season>>
    private set
  lateinit var cast: LiveData<List<CastMember>>
    private set
  lateinit var userComments: LiveData<List<Comment>>
    private set
  lateinit var comments: LiveData<List<Comment>>
    private set
  lateinit var related: LiveData<List<Show>>
    private set
  lateinit var toWatch: LiveData<Episode>
    private set
  lateinit var lastWatched: LiveData<Episode>
    private set
  lateinit var toCollect: LiveData<Episode>
    private set
  lateinit var lastCollected: LiveData<Episode>
    private set

  fun setShowId(showId: Long) {
    if (this.showId == -1L) {
      this.showId = showId

      show = MappedCursorLiveData(
        context,
        Shows.withId(showId),
        ShowMapper.projection,
        null,
        null,
        null,
        ShowMapper
      )
      genres = MappedCursorLiveData(
        context,
        ShowGenres.fromShow(showId),
        arrayOf(ShowGenreColumns.GENRE),
        null,
        null,
        null,
        StringMapper(ShowGenreColumns.GENRE)
      )
      seasons = MappedCursorLiveData(
        context,
        Seasons.fromShow(showId),
        SeasonMapper.projection,
        null,
        null,
        Seasons.DEFAULT_SORT,
        SeasonListMapper
      )
      cast = MappedCursorLiveData(
        context,
        ShowCast.fromShow(showId),
        ShowCastMapper.projection,
        null,
        null,
        Tables.SHOW_CAST + "." + ShowCastColumns.ID + " ASC LIMIT 3",
        ShowCastMapper
      )
      userComments = MappedCursorLiveData(
        context,
        Comments.fromShow(showId),
        CommentMapper.projection,
        CommentColumns.IS_USER_COMMENT + "=1",
        null,
        null,
        CommentListMapper
      )
      comments = MappedCursorLiveData(
        context,
        Comments.fromShow(showId),
        CommentMapper.projection,
        CommentColumns.IS_USER_COMMENT + "=0 AND " + CommentColumns.SPOILER + "=0",
        null,
        CommentColumns.LIKES + " DESC LIMIT 3",
        CommentListMapper
      )
      related = MappedCursorLiveData(
        context,
        RelatedShows.fromShow(showId),
        ShowMapper.projection,
        null,
        null,
        RelatedShowsColumns.RELATED_INDEX + " ASC LIMIT 3",
        ShowListMapper
      )
      toWatch = WatchedLiveData(context, showId)
      lastWatched = MappedCursorLiveData(
        context,
        Episodes.fromShow(showId),
        EpisodeMapper.projection,
        EpisodeColumns.WATCHED + "=1",
        null,
        EpisodeColumns.SEASON + " DESC, " + EpisodeColumns.EPISODE + " DESC LIMIT 1",
        EpisodeMapper
      )
      toCollect = MappedCursorLiveData(
        context,
        Episodes.fromShow(showId),
        EpisodeMapper.projection,
        EpisodeColumns.IN_COLLECTION + "=0 AND " +
            EpisodeColumns.FIRST_AIRED + " IS NOT NULL AND " +
            EpisodeColumns.SEASON + ">0",
        null,
        EpisodeColumns.SEASON + " ASC, " + EpisodeColumns.EPISODE + " ASC LIMIT 1",
        EpisodeMapper
      )
      lastCollected = MappedCursorLiveData(
        context,
        Episodes.fromShow(showId),
        EpisodeMapper.projection,
        EpisodeColumns.IN_COLLECTION + "=1",
        null,
        EpisodeColumns.SEASON + " DESC, " + EpisodeColumns.EPISODE + " DESC LIMIT 1",
        EpisodeMapper
      )

      show.observeForever(showObserver)
    }
  }

  override fun onCleared() {
    show.removeObserver(showObserver)
    super.onCleared()
  }

  private val showObserver = Observer<Show> { show ->
    if (show != null) {
      viewModelScope.launch {
        val currentTime = System.currentTimeMillis()
        val needsSync = show.needsSync
        if (needsSync || currentTime > show.lastSync + SYNC_INTERVAL) {
          syncShow.invokeAsync(SyncShow.Params(show.traktId))
        }

        if (currentTime > show.lastCommentSync + SYNC_INTERVAL_COMMENTS) {
          syncShowComments.invokeAsync(SyncShowComments.Params(show.traktId))
        }

        if (show.lastSync > show.lastCreditsSync) {
          syncShowCredits.invokeAsync(SyncShowCredits.Params(show.traktId))
        }

        if (show.lastSync > show.lastRelatedSync) {
          syncRelatedShows.invokeAsync(SyncRelatedShows.Params(show.traktId))
        }
      }
    }
  }

  override suspend fun onRefresh() {
    val traktId = showHelper.getTraktId(showId)
    val tmdbId = showHelper.getTmdbId(showId)

    val showDeferred = syncShow.invokeAsync(SyncShow.Params(traktId))
    val relatedDeferred = syncRelatedShows.invokeAsync(SyncRelatedShows.Params(traktId))
    val imagesDeferred = syncShowImages.invokeAsync(SyncShowImages.Params(tmdbId))
    val commentsDeferred = syncShowComments.invokeAsync(SyncShowComments.Params(traktId))
    val creditsDeferred = syncShowCredits.invokeAsync(SyncShowCredits.Params(traktId))

    if (TraktLinkSettings.isLinked(context)) {
      val watchedDeferred = syncShowWatchedStatus.invokeAsync(SyncShowWatchedStatus.Params(traktId))
      val collectedDeferred =
        syncShowCollectedStatus.invokeAsync(SyncShowCollectedStatus.Params(traktId))
      watchedDeferred.await()
      collectedDeferred.await()
    }

    showDeferred.await()
    relatedDeferred.await()
    imagesDeferred.await()
    commentsDeferred.await()
    creditsDeferred.await()
    return
  }

  companion object {

    private const val SYNC_INTERVAL = 2 * DateUtils.DAY_IN_MILLIS
    private const val SYNC_INTERVAL_COMMENTS = 3 * DateUtils.HOUR_IN_MILLIS
  }
}
